File: typescript/base-hmi-types/src/projects/HmiProjectFolder.ts
import { HmiProjectFolderType } from "./HmiProjectFolderType.js";
import { IHmiProjectFolder } from "./IHmiProject.js";
import { IHmiProjectItemDescriptor } from "./HmiProjectItemDescriptor.js";

export class HmiProjectFolder implements IHmiProjectFolder {
  private readonly getFoldersCore: (signal?: AbortSignal) => Promise<readonly IHmiProjectFolder[]>;
  private readonly getItemsCore: (signal?: AbortSignal) => Promise<readonly IHmiProjectItemDescriptor[]>;
  private folderValues: readonly IHmiProjectFolder[];
  private itemValues: readonly IHmiProjectItemDescriptor[];

  readonly id?: string | undefined;
  readonly name: string;
  readonly path: string;
  readonly folderType: HmiProjectFolderType;

  constructor(
    folderType: HmiProjectFolderType,
    name: string,
    id?: string,
    path?: string,
    getFolders?:
      | readonly IHmiProjectFolder[]
      | ((signal?: AbortSignal) => Promise<readonly IHmiProjectFolder[]>),
    getItems?:
      | readonly IHmiProjectItemDescriptor[]
      | ((signal?: AbortSignal) => Promise<readonly IHmiProjectItemDescriptor[]>),
  ) {
    this.folderType = folderType;
    this.name = name;
    this.id = id;
    this.path = path ?? name;
    this.folderValues = typeof getFolders === "function" ? [] : (getFolders ?? []);
    this.itemValues = typeof getItems === "function" ? [] : (getItems ?? []);
    this.getFoldersCore =
      typeof getFolders === "function"
        ? async signal => (this.folderValues = await getFolders(signal))
        : () => Promise.resolve(this.folderValues);
    this.getItemsCore =
      typeof getItems === "function"
        ? async signal => (this.itemValues = await getItems(signal))
        : () => Promise.resolve(this.itemValues);
  }

  get folders(): readonly IHmiProjectFolder[] {
    return this.folderValues;
  }

  get items(): readonly IHmiProjectItemDescriptor[] {
    return this.itemValues;
  }

  getFolders(signal?: AbortSignal): Promise<readonly IHmiProjectFolder[]> {
    return this.getFoldersCore(signal);
  }

  getItems(signal?: AbortSignal): Promise<readonly IHmiProjectItemDescriptor[]> {
    return this.getItemsCore(signal);
  }
}

File: src/BaseHmiTypes/Projects/HmiProjectFolder.cs
namespace BaseHmiTypes.Projects;

public class HmiProjectFolder : IHmiProjectFolder
{
    private readonly Func<CancellationToken, ValueTask<IReadOnlyList<IHmiProjectFolder>>> getFolders;
    private readonly Func<CancellationToken, ValueTask<IReadOnlyList<IHmiProjectItemDescriptor>>> getItems;

    public HmiProjectFolder(
        HmiProjectFolderType folderType,
        string name,
        string? id = null,
        string? path = null,
        Func<CancellationToken, ValueTask<IReadOnlyList<IHmiProjectFolder>>>? getFolders = null,
        Func<CancellationToken, ValueTask<IReadOnlyList<IHmiProjectItemDescriptor>>>? getItems = null)
    {
        FolderType = folderType;
        Name = name;
        Id = id;
        Path = path ?? name;
        this.getFolders = getFolders ?? (_ => new ValueTask<IReadOnlyList<IHmiProjectFolder>>(Array.Empty<IHmiProjectFolder>()));
        this.getItems = getItems ?? (_ => new ValueTask<IReadOnlyList<IHmiProjectItemDescriptor>>(Array.Empty<IHmiProjectItemDescriptor>()));
    }

    public string? Id { get; }

    public string Name { get; }

    public string Path { get; }

    public HmiProjectFolderType FolderType { get; }

    public IReadOnlyList<IHmiProjectFolder> Folders => GetFoldersAsync().AsTask().GetAwaiter().GetResult();

    public IReadOnlyList<IHmiProjectItemDescriptor> Items => GetItemsAsync().AsTask().GetAwaiter().GetResult();

    public ValueTask<IReadOnlyList<IHmiProjectFolder>> GetFoldersAsync(CancellationToken cancellationToken = default)
    {
        return getFolders(cancellationToken);
    }

    public ValueTask<IReadOnlyList<IHmiProjectItemDescriptor>> GetItemsAsync(CancellationToken cancellationToken = default)
    {
        return getItems(cancellationToken);
    }
}
